![logo](images/untumbes.PNG)

<center><b>Prof. Dr. Jorge Zavaleta - zavaleta.jorge@gmail.com</b></center>

# Algoritmo FP-Growth

In [1]:
#librarys
import pandas as pd
import numpy as np
import random
#
import heapq
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import fpgrowth
#
from surprise import Dataset, Reader, KNNBasic
# ml
from surprise.model_selection import train_test_split

# graphics
import matplotlib.pylab as plt
%matplotlib inline
#
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Passo 1: Gerar um conjunto de dados aleatório
data = [['Leite', 'Cebola', 'Noz-Moscada', 'Feijão', 'Ovos', 'Iogurte'],
        ['Dill', 'Cebola', 'Noz-Moscada', 'Feijão', 'Ovos', 'Iogurte'],
        ['Leite', 'Maçã', 'Feijão', 'Ovos'],
        ['Leite', 'Milho', 'Noz-Moscada', 'Feijão', 'Iogurte'],
        ['Milho', 'Cebola', 'Cebola', 'Feijão', 'Sorvete', 'Ovos']]
data

[['Leite', 'Cebola', 'Noz-Moscada', 'Feijão', 'Ovos', 'Iogurte'],
 ['Dill', 'Cebola', 'Noz-Moscada', 'Feijão', 'Ovos', 'Iogurte'],
 ['Leite', 'Maçã', 'Feijão', 'Ovos'],
 ['Leite', 'Milho', 'Noz-Moscada', 'Feijão', 'Iogurte'],
 ['Milho', 'Cebola', 'Cebola', 'Feijão', 'Sorvete', 'Ovos']]

In [3]:
# Passo 2: Pré-processamento dos dados
te = TransactionEncoder()
te_tmp = te.fit(data).transform(data)
df = pd.DataFrame(te_tmp, columns=te.columns_)
df

Cebola   Dill  Feijão  Iogurte  Leite   Maçã  Milho  Noz-Moscada   Ovos  \
0    True  False    True     True   True  False  False         True   True   
1    True   True    True     True  False  False  False         True   True   
2   False  False    True    False   True   True  False        False   True   
3   False  False    True     True   True  False   True         True  False   
4    True  False    True    False  False  False   True        False   True   

   Sorvete  
0    False  
1    False  
2    False  
3    False  
4     True

In [4]:
df.shape

(5, 10)

In [5]:
# Passo 3: Aplicar o algoritmo FP-Growth
frequent_itemsets = fpgrowth(df, min_support=0.6, use_colnames=True)
frequent_itemsets[['itemsets','support']]

itemsets  support
0                         (Feijão)      1.0
1                           (Ovos)      0.8
2                    (Noz-Moscada)      0.6
3                          (Leite)      0.6
4                        (Iogurte)      0.6
5                         (Cebola)      0.6
6                   (Ovos, Feijão)      0.8
7            (Feijão, Noz-Moscada)      0.6
8                  (Leite, Feijão)      0.6
9           (Noz-Moscada, Iogurte)      0.6
10               (Feijão, Iogurte)      0.6
11  (Feijão, Noz-Moscada, Iogurte)      0.6
12                  (Cebola, Ovos)      0.6
13                (Cebola, Feijão)      0.6
14          (Cebola, Ovos, Feijão)      0.6

In [6]:
# Collecting the inferred rules in a dataframe 
reglas = association_rules(frequent_itemsets, metric ="lift", min_threshold = 1) 
reglas = reglas.sort_values(['confidence', 'lift'], ascending =[False, False]) 
reglas.head(10)

antecedents            consequents  antecedent support  \
6           (Noz-Moscada)              (Iogurte)                 0.6   
7               (Iogurte)          (Noz-Moscada)                 0.6   
10  (Feijão, Noz-Moscada)              (Iogurte)                 0.6   
11      (Feijão, Iogurte)          (Noz-Moscada)                 0.6   
14          (Noz-Moscada)      (Feijão, Iogurte)                 0.6   
15              (Iogurte)  (Feijão, Noz-Moscada)                 0.6   
16               (Cebola)                 (Ovos)                 0.6   
21       (Cebola, Feijão)                 (Ovos)                 0.6   
23               (Cebola)         (Ovos, Feijão)                 0.6   
0                  (Ovos)               (Feijão)                 0.8   

    consequent support  support  confidence      lift  leverage  conviction  \
6                  0.6      0.6         1.0  1.666667      0.24         inf   
7                  0.6      0.6         1.0  1.666667      0.24         inf   
10                 0.6      0.6         1.0  1.666667      0.24         inf   
11                 0.6      0.6         1.0  1.666667      0.24         inf   
14                 0.6      0.6         1.0  1.666667      0.24         inf   
15                 0.6      0.6         1.0  1.666667      0.24         inf   
16                 0.8      0.6         1.0  1.250000      0.12         inf   
21                 0.8      0.6         1.0  1.250000      0.12         inf   
23                 0.8      0.6         1.0  1.250000      0.12         inf   
0                  1.0      0.8         1.0  1.000000      0.00         inf   

    zhangs_metric  
6             1.0  
7             1.0  
10            1.0  
11            1.0  
14            1.0  
15            1.0  
16            0.5  
21            0.5  
23            0.5  
0             0.0

>## Análisis

Podemos interpretar las reglas resultantes de la siguiente manera:

Valor de soporte -> muestra que quien compra Noz-Moscada compra Iogurte tiene un soporte de 0.6

Valor de Confianza -> muestra que el 100% de los clientes que compran Noz-Moscada también compran Iogurte.

Valor de elevación -> muestra que las ventas de (Iogurte) aumentan en 1.67 para las compras con Noz-Moscada.

>## Cesta do Tio José

In [16]:
# Load and preprocess data set
ctjose = pd.read_csv('data/cesta.csv', sep=';', encoding='latin-1')
ctjose.set_index('TID', inplace=True)
ctjose.head(10)

Café  Desinfetante  Carne suina  Detergente  Limpador multiuso  \
TID                                                                   
0       1             0            1           0                  1   
1       1             0            1           0                  1   
2       0             1            0           1                  0   
3       1             0            1           0                  1   
4       0             1            0           1                  0   
5       1             0            1           0                  1   
6       0             1            0           1                  0   
7       1             0            1           0                  1   
8       0             1            0           1                  0   
9       1             0            1           0                  1   

     Limpa pisos de madeira  Frango  agua  Esponjas  Limpa-vidros  ...  \
TID                                                                ...   
0                         0       1     0         1             0  ...   
1                         0       1     0         1             0  ...   
2                         1       0     1         0             1  ...   
3                         0       1     0         1             0  ...   
4                         1       0     1         0             1  ...   
5                         0       1     0         1             0  ...   
6                         1       0     1         0             1  ...   
7                         0       1     0         1             0  ...   
8                         1       0     1         0             1  ...   
9                         0       1     0         1             0  ...   

     Baldes  Tomates  Leite  alcool  Panos de chao  Carne bovina  Cha  \
TID                                                                     
0         0        1      0       1              0             1    0   
1         0        1      0       1              0             1    0   
2         1        0      1       0              1             0    1   
3         0        1      0       1              0             1    0   
4         1        0      1       0              1             0    1   
5         0        1      0       1              0             1    0   
6         1        0      1       0              1             0    1   
7         0        1      0       1              0             1    0   
8         1        0      1       0              1             0    1   
9         0        1      0       1              0             1    0   

     Bananas  Suco de laranja  Carne de peixe  
TID                                            
0          1                0               1  
1          1                0               1  
2          0                1               0  
3          1                0               1  
4          0                1               0  
5          1                0               1  
6          0                0               1  
7          1                0               1  
8          0                1               0  
9          1                0               1  

[10 rows x 25 columns]

In [17]:
ctjose.shape

(23, 25)

In [20]:
# Passo 2: Pré-processamento dos dados- ya está en 1/0
#te_tj = TransactionEncoder()
#te_tj_tmp = te_tj.fit(ctjose).transform(ctjose)
#data_tj = pd.DataFrame(te_tj_tmp, columns=te_tj.columns_)
#data_tj.head()

In [24]:
# aplicando el algoritmo
cesta_items = fpgrowth(ctjose, min_support=0.5, use_colnames=True)
cesta_items[['itemsets','support']]

itemsets   support
0                                      (Carne de peixe)  0.608696
1                                             (Bananas)  0.565217
2                                        (Carne bovina)  0.565217
3                                              (alcool)  0.521739
4                                             (Tomates)  0.521739
...                                                 ...       ...
8186  (Sabao neutro ou de coco, Laranjas, Tomates, C...  0.521739
8187  (Sabao neutro ou de coco, Laranjas, Tomates, C...  0.521739
8188  (Sabao neutro ou de coco, Laranjas, Tomates, C...  0.521739
8189  (Sabao neutro ou de coco, Laranjas, Tomates, a...  0.521739
8190  (Sabao neutro ou de coco, Laranjas, Tomates, C...  0.521739

[8191 rows x 2 columns]

In [25]:
# Collecting the inferred rules in a dataframe 
tjose_rules = association_rules(cesta_items, metric ="confidence", min_threshold = 0.5) 
tjose_rules = tjose_rules.sort_values(['confidence', 'lift'], ascending =[False, False]) 
tjose_rules.head(10) 

antecedents                consequents  antecedent support  \
66                   (alcool)                  (Tomates)            0.521739   
67                  (Tomates)                   (alcool)            0.521739   
68                   (alcool)                 (Laranjas)            0.521739   
69                 (Laranjas)                   (alcool)            0.521739   
70                   (alcool)                    (Macas)            0.521739   
71                    (Macas)                   (alcool)            0.521739   
72                   (alcool)  (Sabao neutro ou de coco)            0.521739   
73  (Sabao neutro ou de coco)                   (alcool)            0.521739   
74                   (alcool)                 (Esponjas)            0.521739   
75                 (Esponjas)                   (alcool)            0.521739   

    consequent support   support  confidence      lift  leverage  conviction  \
66            0.521739  0.521739         1.0  1.916667  0.249527         inf   
67            0.521739  0.521739         1.0  1.916667  0.249527         inf   
68            0.521739  0.521739         1.0  1.916667  0.249527         inf   
69            0.521739  0.521739         1.0  1.916667  0.249527         inf   
70            0.521739  0.521739         1.0  1.916667  0.249527         inf   
71            0.521739  0.521739         1.0  1.916667  0.249527         inf   
72            0.521739  0.521739         1.0  1.916667  0.249527         inf   
73            0.521739  0.521739         1.0  1.916667  0.249527         inf   
74            0.521739  0.521739         1.0  1.916667  0.249527         inf   
75            0.521739  0.521739         1.0  1.916667  0.249527         inf   

    zhangs_metric  
66            1.0  
67            1.0  
68            1.0  
69            1.0  
70            1.0  
71            1.0  
72            1.0  
73            1.0  
74            1.0  
75            1.0

## Análisis

Podemos interpretar las reglas resultantes de la siguiente manera:

Valor de soporte -> muestra que quien compra Álcool compra tomates tiene un soporte de 0.52

Valor de Confianza -> muestra que el 100% de los clientes que compran {Álcool,Tomates,Laranjas,Macas} también compran {Tomates,alcool,laranjas, Macas}.

Valor de elevación -> muestra que las ventas de (Tomates) aumentan en 1.92 para las compras con Alcool.

---
<center><b>&copy;Jorge Zavaleta, 2024</b></center>